File: ios/SakanHub/SakanHub/Core/Config/FeatureFlags.swift
import Foundation

/// Feature flags for the "ادخل العقار" (Enter Property) subsystem.
///
/// All flags DEFAULT to OFF. Reading from Info.plist means a build
/// (via `Sakan.xcconfig`) enables them; there is no runtime UI to
/// toggle them, and no server round-trip that could turn them on
/// unexpectedly.
///
/// Info.plist keys (String, "YES" / "NO"):
///   - PROPERTY_CAMERA_TEST_ENABLED
///   - PROPERTY_3D_CAPTURE_ENABLED
///   - PROPERTY_3D_VIEWER_ENABLED
///   - PROPERTY_AR_ENABLED
///
/// ⚠️  Enabling PROPERTY_3D_CAPTURE_ENABLED simply exposes the capture
///     UI. There is NO real 3D Gaussian Splatting pipeline; the backend
///     will respond with `processingStatus = failed` + "not configured"
///     until a real GPU service is wired in. The UI must surface that
///     message honestly and NOT show a "your 3D tour is ready" state.
enum FeatureFlags {

    private static func read(_ key: String) -> Bool {
        guard let raw = Bundle.main.object(forInfoDictionaryKey: key) as? String else {
            return false
        }
        return (raw as NSString).boolValue
    }

    static var propertyCameraTestEnabled: Bool { read("PROPERTY_CAMERA_TEST_ENABLED") }
    static var property3DCaptureEnabled:  Bool { read("PROPERTY_3D_CAPTURE_ENABLED")  }
    static var property3DViewerEnabled:   Bool { read("PROPERTY_3D_VIEWER_ENABLED")   }
    static var propertyAREnabled:         Bool { read("PROPERTY_AR_ENABLED")          }

    /// Any capture / viewer / AR entry point.
    static var enterPropertyAnyEnabled: Bool {
        propertyCameraTestEnabled
            || property3DCaptureEnabled
            || property3DViewerEnabled
            || propertyAREnabled
    }
}

File: backend/src/config/feature-flags.ts
/**
 * Feature flags — "Enter Property" (ادخل العقار) subsystem.
 *
 * ⚠️  All flags default to FALSE. Enabling any flag simply exposes a route
 *     or UI entry; it does NOT create infrastructure. In particular:
 *
 *     - PROPERTY_3D_CAPTURE_ENABLED does NOT provide a working 3D
 *       Gaussian Splatting pipeline — that requires an external GPU
 *       service which is not part of this repository.
 *     - PROPERTY_AR_ENABLED gates iOS AR entry points only.
 *
 * Reading env at module load is deliberate: the flags are static per
 * process, and tests can override by mutating `featureFlags` directly.
 */

function parseBool(v: string | undefined): boolean {
  if (!v) return false;
  const s = v.trim().toLowerCase();
  return s === '1' || s === 'true' || s === 'yes' || s === 'on';
}

export interface FeatureFlags {
  PROPERTY_CAMERA_TEST_ENABLED: boolean;
  PROPERTY_3D_CAPTURE_ENABLED:  boolean;
  PROPERTY_3D_VIEWER_ENABLED:   boolean;
  PROPERTY_AR_ENABLED:          boolean;
}

export const featureFlags: FeatureFlags = {
  PROPERTY_CAMERA_TEST_ENABLED: parseBool(process.env.PROPERTY_CAMERA_TEST_ENABLED),
  PROPERTY_3D_CAPTURE_ENABLED:  parseBool(process.env.PROPERTY_3D_CAPTURE_ENABLED),
  PROPERTY_3D_VIEWER_ENABLED:   parseBool(process.env.PROPERTY_3D_VIEWER_ENABLED),
  PROPERTY_AR_ENABLED:          parseBool(process.env.PROPERTY_AR_ENABLED),
};

export function isFlagEnabled(name: keyof FeatureFlags): boolean {
  return featureFlags[name] === true;
}
